HCS design document

steven-sheehy opened the pull request:
**Detailed description**:
Initial design document for HCS

**Which issue(s) this PR fixes**:
Fixes #327 

**Special notes for your reviewer**:
Will change base branch once HCS branch is cut

**Checklist**
- [x] Documentation added
- [ ] Tests updated

Nana-EC commented on docs/design/hcs.md:
```diff
@@ -0,0 +1,150 @@
+}
+
+message ConsensusTopicResponse {
+    .proto.Timestamp consensusTimestamp = 1; // The time at which the transaction reach consensus
```
nit: *reached

steven-sheehy replied:
Thanks, done

Nana-EC commented on docs/design/hcs.md:
```diff
@@ -0,0 +1,150 @@
+
+message ConsensusTopicResponse {
+    .proto.Timestamp consensusTimestamp = 1; // The time at which the transaction reach consensus
+    bytes message = 2; // The message body originally in the ConsensusSubmitMessageTransactionBody
```
is there a potential size limit on this? If so what is it?

steven-sheehy replied:
Hedera imposes a 4K maximum proto request size, so the message is less than 4K and varies depending upon the variable values in the original request. I will add a note here about the 4K.

steven-sheehy replied:
Done

Nana-EC commented on docs/design/hcs.md:
```diff
@@ -0,0 +1,150 @@
+- Create a Systemd unit file
+- Add to CircleCI and deployment script
+
+## Parser
```
missing details under Parser. Again unsure if this is a place holder or not

steven-sheehy replied:
Intentional grouping of sections

steven-sheehy commented on docs/design/hcs.md:
```diff
@@ -0,0 +1,150 @@
+# Hedera Consensus Service Design
```
On the server side, `subscribeTopic` is a void return type. The StreamObserver pushes the objects mentioned in the protobuf to the client as they come in. All the interactions are reactive, so they're a lot harder to show in a sequence diagram as they're not sequential. :)

steven-sheehy replied:
Done

steven-sheehy commented on docs/design/hcs.md:
```diff
@@ -0,0 +1,150 @@
+4. GRPC API retrieves the current messages from PostgreSQL and returns to client
+5. GRPC API is notified of new `ConsensusSubmitMessage` transactions via PostgreSQL Notify
+
+## GRPC API
```
That's just a H2 section header to group all GRPC sections. The protobuf (H3) is the first content

steven-sheehy commented:
> Shouldn't we return the 'running hash' as well?
This is still being debated about whether it is included in the HAPI layer. I took it out after Mike said he was taking it out of HAPI, but still up in the air.

> The returns from StreamObserver to the clients are of type 'consensusTypeResponse', right?

Yes, I will update the diagram to show more types.

> Is there a need for a peek() or areThereUnreadMessagesInTheTopic() kind of API?

Unread by whom? Multiple different clients could request messages for the same topic ID, so tracking client state would be difficult/not scalable. The clients should track their unread state and can resubmit request with startTime=prevStartTime+1.

> For scaling, will there be multiple "horizontally scalable" instances of StreamObserver? If so, we should somehow make it clear

`StreamObserver` is a part of the GRPC API and is managed by it, so there will be one per `subscribeTopic()` call. I could technically remove it from the diagram and just show calls directly back to client, but wanted to show the thing shared between listener and repository to send back.

> Do we see a need for a getTopicInfo() API at the mirror node, or would something like that be handled purely on the HCS service side? If we do (say to return some statistics about number of messages, etc.), then we might need CONSENSUSREADTOPIC too.

We can always add APIs as the need arises (like this or websockets), but would like to keep this initial phase simpler so we can get something out the door.

> GRPC streaming interface has error codes like DEADLINE_EXCEEDED. Do we need to define/implement those?

`io.grpc.Status` already contains that error code and a suite of others already defined that we should utilize to provide common error handling across clients. We can probably augment it with any additional ones that we need, but should stick to the built in ones if possible with a helpful message alongside it:
```java
streamObserver.onError(Status.DEADLINE_EXCEEDED.augmentDescription(e.getMessage()).withCause(e).asRuntimeException());
```

mike-burrage-hedera commented on docs/design/hcs.md:
```diff
@@ -0,0 +1,150 @@
+2. Parser persists to PostgreSQL
+3. Client queries GRPC API for messages from a specific topic
+4. GRPC API retrieves the current messages from PostgreSQL and returns to client
+5. GRPC API is notified of new `ConsensusSubmitMessage` transactions via PostgreSQL Notify
```
For completeness sake, the GRPC API should be notified of ConsensusUpdateTopic and ConsensusDeleteTopic messages so as to know when to close a topic.

steven-sheehy replied:
I'm thinking the client should contain the read state like Kafka. There can be multiple clients that want to subscribe to the same topic and managing their state would be difficult. If that is the case, the GRPC API doesn't need notification of update or deletes.

mike-burrage-hedera replied:
I was just thinking that yes the GRPC interface doesn't need to expose those transactions, but the GRPC server can close connections when the topic is closed. Guess it's just an optimization, and not necessary for MVP.

mike-burrage-hedera commented on docs/design/hcs.md:
````diff
@@ -0,0 +1,150 @@
+```postgres-sql
+create table if not exists topic_message (
+  consensus_timestamp bigint primary key not null,
+  entity_id bigint not null references t_entities (id) on delete cascade on update cascade,
````
Use psql domain `nanos_timestamp` instead of bigint on consensus_timestamp.
Instead of `entity_id` use `topic_realm_num realm_num` and `topic_num entity_num`.

steven-sheehy replied:
Done

steven-sheehy commented:
I've updated the document to reflect all comments. I've also updated it to introduce a `TopicMessageService` service layer to provide reuse between different protocols. If we require a 30 min call to discuss the changes, please let me know. Please provide your final feedback by EOD Wednesday.

Nana-EC commented on docs/design/hcs.md:
```diff
@@ -0,0 +1,241 @@
+
+### Persist
+
+Modify `RecordFileLogger` to handle HCS transactions
```
RecordFileLogger is quite bloated already. Is this
1. The right place to add to it?
2. An opportunity to modularize a bit more and make RecordFileLogger cleaner, simpler and easier to test?

steven-sheehy replied:
1. Yes, it's the only place that inserts into `t_transactions`
2. That's what the `Notify` section entails, though that is the lowest priority task. For MVP, we need to insert into the DB with minimal changes. Once that is complete, refactoring can take place as part of the `Notify` task if time permits.

Nana-EC commented on docs/design/hcs.md:
```diff
@@ -0,0 +1,241 @@
+Modify `RecordFileLogger` to handle HCS transactions
+
+-   Modify Entities to handle Topic ID and new fields
+-   Insert HCS transaction into `t_transactions`
```
Will this be done synchronously or asynchronously with the current db updates based on record file streams?

steven-sheehy replied:
If I understand you correctly, it will be done synchronously the same way other transactions are inserted into `t_transactions`.

gregscullard commented on docs/design/hcs.md:
```diff
@@ -0,0 +1,241 @@
+    uint64 limit = 4; // The maximum number of messages to receive before stopping. If not set or set to zero it will return messages indefinitely.
+}
+
+message ConsensusTopicResponse {
```
This should (optionally ?) include the record for the `consensusMessage` ?
-Unless mirror node only notifies of `SUCCESS`ful messages on topic, the consumer has no means of knowing whether the message was successful or not.
-The submitter would know if the transaction was successful (providing they wait for a receipt), but any consumer other than the submitter would not.

mike-burrage-hedera replied:
Only successful messages get handled by the mirror-node and published out.

gregscullard commented on docs/design/hcs.md:
```diff
@@ -0,0 +1,241 @@
+
+message ConsensusTopicResponse {
+    .proto.Timestamp consensusTimestamp = 1; // The time at which the transaction reached consensus
+    bytes message = 2; // The message body originally in the ConsensusSubmitMessageTransactionBody. Message size will be less than 4K.
```
Suggest the full `ConsensusSubmitMessageTransaction` is included here, this will be necessary when state proofs are available (state proof is on whole transaction, not just contents thereof), this would avoid having to refactor the message in the future.
This may also be useful so the subscriber knows who originated the transaction.

steven-sheehy replied:
As discussed on the call, we will keep `subscribeTopic()` as is and potentially add another generic API with the full transaction and record in another release.
